prov/tcp/src/tcpx_progress.c

```diff
@@ -418,7 +428,12 @@ static int tcpx_handle_ack(struct tcpx_ep *ep)
 	    sizeof(ep->cur_rx.hdr.base_hdr))
 		return -FI_EIO;
 
-	assert(!slist_empty(&ep->need_ack_queue));
+	if (slist_empty(&ep->need_ack_queue))
+	{
+		/* Early response */
+		return -FI_EAGAIN;
+	}
```

Comment:
I think what we can do here is:
```
if (slist_empty(&ep->need_ack_queue)) {
    assert(ep->cur_tx->flags & TCPX_NEED_ACK)
    ep->cur_tx->flags &= ~TCPX_NEED_ACK;
    return FI_SUCCESS;
}
```
No other changes for this patch should be necessary, I think.

Reply:
Thanks, I will update the patch with your suggestion.